File: concept/thing/AttributeImpl.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.concept.thing;

import com.eclipsesource.json.Json;
import com.eclipsesource.json.JsonObject;
import com.eclipsesource.json.JsonValue;
import com.vaticle.typedb.client.api.TypeDBTransaction;
import com.vaticle.typedb.client.api.concept.thing.Attribute;
import com.vaticle.typedb.client.api.concept.type.ThingType;
import com.vaticle.typedb.client.common.exception.TypeDBClientException;
import com.vaticle.typedb.client.concept.type.AttributeTypeImpl;
import com.vaticle.typedb.protocol.ConceptProto;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.stream.Stream;

import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.BAD_VALUE_TYPE;
import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.INVALID_CONCEPT_CASTING;
import static com.vaticle.typedb.client.common.rpc.RequestBuilder.Thing.Attribute.getOwnersReq;
import static com.vaticle.typedb.client.concept.type.ThingTypeImpl.protoThingType;
import static com.vaticle.typedb.common.collection.Bytes.bytesToHexString;
import static com.vaticle.typedb.common.util.Objects.className;

public abstract class AttributeImpl<VALUE> extends ThingImpl implements Attribute<VALUE> {
    AttributeImpl(java.lang.String iid, boolean isInferred) {
        super(iid, isInferred);
    }

    public static AttributeImpl<?> of(ConceptProto.Thing thingProto) {
        switch (thingProto.getType().getValueType()) {
            case BOOLEAN:
                return AttributeImpl.Boolean.of(thingProto);
            case LONG:
                return AttributeImpl.Long.of(thingProto);
            case DOUBLE:
                return AttributeImpl.Double.of(thingProto);
            case STRING:
                return AttributeImpl.String.of(thingProto);
            case DATETIME:
                return AttributeImpl.DateTime.of(thingProto);
            case UNRECOGNIZED:
            default:
                throw new TypeDBClientException(BAD_VALUE_TYPE, thingProto.getType().getValueType());
        }
    }

    @Override
    public abstract AttributeTypeImpl getType();

    @Override
    public AttributeImpl<VALUE> asAttribute() {
        return this;
    }

    @Override
    public AttributeImpl.Boolean asBoolean() {
        throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Attribute.Boolean.class));
    }

    @Override
    public AttributeImpl.Long asLong() {
        throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Attribute.Long.class));
    }

    @Override
    public AttributeImpl.Double asDouble() {
        throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Attribute.Double.class));
    }

    @Override
    public AttributeImpl.String asString() {
        throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Attribute.String.class));
    }

    @Override
    public AttributeImpl.DateTime asDateTime() {
        throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Attribute.DateTime.class));
    }

    public abstract VALUE getValue();

    public abstract static class Remote<VALUE> extends ThingImpl.Remote implements Attribute.Remote<VALUE> {

        Remote(TypeDBTransaction transaction, java.lang.String iid, boolean isInferred) {
            super(transaction, iid, isInferred);
        }

        @Override
        public final Stream<ThingImpl> getOwners() {
            return stream(getOwnersReq(getIID()))
                    .flatMap(rp -> rp.getAttributeGetOwnersResPart().getThingsList().stream())
                    .map(ThingImpl::of);
        }

        @Override
        public Stream<ThingImpl> getOwners(ThingType ownerType) {
            return stream(getOwnersReq(getIID(), protoThingType(ownerType)))
                    .flatMap(rp -> rp.getAttributeGetOwnersResPart().getThingsList().stream())
                    .map(ThingImpl::of);
        }

        @Override
        public abstract AttributeTypeImpl getType();

        @Override
        public AttributeImpl.Remote<VALUE> asAttribute() {
            return this;
        }

        @Override
        public AttributeImpl.Boolean.Remote asBoolean() {
            throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Attribute.Boolean.class));
        }

        @Override
        public AttributeImpl.Long.Remote asLong() {
            throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Attribute.Long.class));
        }

        @Override
        public AttributeImpl.Double.Remote asDouble() {
            throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Attribute.Double.class));
        }

        @Override
        public AttributeImpl.String.Remote asString() {
            throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Attribute.String.class));
        }

        @Override
        public AttributeImpl.DateTime.Remote asDateTime() {
            throw new TypeDBClientException(INVALID_CONCEPT_CASTING, className(this.getClass()), className(Attribute.DateTime.class));
        }

        public abstract VALUE getValue();
    }

    public static class Boolean extends AttributeImpl<java.lang.Boolean> implements Attribute.Boolean {

        private final AttributeTypeImpl.Boolean type;
        private final java.lang.Boolean value;

        Boolean(java.lang.String iid, boolean isInferred, AttributeTypeImpl.Boolean type, boolean value) {
            super(iid, isInferred);
            this.type = type;
            this.value = value;
        }

        public static AttributeImpl.Boolean of(ConceptProto.Thing thingProto) {
            return new AttributeImpl.Boolean(
                    bytesToHexString(thingProto.getIid().toByteArray()),
                    thingProto.getInferred(),
                    AttributeTypeImpl.Boolean.of(thingProto.getType()),
                    thingProto.getValue().getBoolean()
            );
        }

        @Override
        public final AttributeTypeImpl.Boolean getType() {
            return type;
        }

        @Override
        public final java.lang.Boolean getValue() {
            return value;
        }

        @Override
        public final AttributeImpl.Boolean asBoolean() {
            return this;
        }

        @Override
        public AttributeImpl.Boolean.Remote asRemote(TypeDBTransaction transaction) {
            return new AttributeImpl.Boolean.Remote(transaction, getIID(), isInferred(), type, value);
        }

        public static class Remote extends AttributeImpl.Remote<java.lang.Boolean> implements Attribute.Boolean.Remote {

            private final AttributeTypeImpl.Boolean type;
            private final java.lang.Boolean value;

            Remote(TypeDBTransaction transaction, java.lang.String iid, boolean isInferred, AttributeTypeImpl.Boolean type, java.lang.Boolean value) {
                super(transaction, iid, isInferred);
                this.type = type;
                this.value = value;
            }

            @Override
            public Attribute.Boolean.Remote asRemote(TypeDBTransaction transaction) {
                return new AttributeImpl.Boolean.Remote(transaction, getIID(), isInferred(), type, value);
            }

            @Override
            public final java.lang.Boolean getValue() {
                return value;
            }

            @Override
            public AttributeTypeImpl.Boolean getType() {
                return type;
            }

            @Override
            public final AttributeImpl.Boolean.Remote asBoolean() {
                return this;
            }
        }
    }

    public static class Long extends AttributeImpl<java.lang.Long> implements Attribute.Long {

        private final AttributeTypeImpl.Long type;
        private final long value;

        Long(java.lang.String iid, boolean isInferred, AttributeTypeImpl.Long type, long value) {
            super(iid, isInferred);
            this.type = type;
            this.value = value;
        }

        public static AttributeImpl.Long of(ConceptProto.Thing thingProto) {
            return new AttributeImpl.Long(
                    bytesToHexString(thingProto.getIid().toByteArray()),
                    thingProto.getInferred(),
                    AttributeTypeImpl.Long.of(thingProto.getType()),
                    thingProto.getValue().getLong()
            );
        }

        @Override
        public final AttributeTypeImpl.Long getType() {
            return type;
        }

        @Override
        public final java.lang.Long getValue() {
            return value;
        }

        @Override
        public final AttributeImpl.Long asLong() {
            return this;
        }

        @Override
        public AttributeImpl.Long.Remote asRemote(TypeDBTransaction transaction) {
            return new AttributeImpl.Long.Remote(transaction, getIID(), isInferred(), type, value);
        }

        public static class Remote extends AttributeImpl.Remote<java.lang.Long> implements Attribute.Long.Remote {

            private final AttributeTypeImpl.Long type;
            private final long value;

            Remote(TypeDBTransaction transaction, java.lang.String iid, boolean isInferred, AttributeTypeImpl.Long type, long value) {
                super(transaction, iid, isInferred);
                this.type = type;
                this.value = value;
            }

            @Override
            public Attribute.Long.Remote asRemote(TypeDBTransaction transaction) {
                return new AttributeImpl.Long.Remote(transaction, getIID(), isInferred(), type, value);
            }

            @Override
            public final java.lang.Long getValue() {
                return value;
            }

            @Override
            public AttributeTypeImpl.Long getType() {
                return type;
            }

            @Override
            public final AttributeImpl.Long.Remote asLong() {
                return this;
            }
        }
    }

    public static class Double extends AttributeImpl<java.lang.Double> implements Attribute.Double {

        private final AttributeTypeImpl.Double type;
        private final double value;

        Double(java.lang.String iid, boolean isInferred, AttributeTypeImpl.Double type, double value) {
            super(iid, isInferred);
            this.type = type;
            this.value = value;
        }

        public static AttributeImpl.Double of(ConceptProto.Thing thingProto) {
            return new AttributeImpl.Double(
                    bytesToHexString(thingProto.getIid().toByteArray()),
                    thingProto.getInferred(),
                    AttributeTypeImpl.Double.of(thingProto.getType()),
                    thingProto.getValue().getDouble()
            );
        }

        @Override
        public final AttributeTypeImpl.Double getType() {
            return type;
        }

        @Override
        public final java.lang.Double getValue() {
            return value;
        }

        @Override
        public final AttributeImpl.Double asDouble() {
            return this;
        }

        @Override
        public AttributeImpl.Double.Remote asRemote(TypeDBTransaction transaction) {
            return new AttributeImpl.Double.Remote(transaction, getIID(), isInferred(), type, value);
        }

        public static class Remote extends AttributeImpl.Remote<java.lang.Double> implements Attribute.Double.Remote {

            private final AttributeTypeImpl.Double type;
            private final double value;

            Remote(TypeDBTransaction transaction, java.lang.String iid, boolean isInferred, AttributeTypeImpl.Double type, double value) {
                super(transaction, iid, isInferred);
                this.type = type;
                this.value = value;
            }

            @Override
            public Attribute.Double.Remote asRemote(TypeDBTransaction transaction) {
                return new AttributeImpl.Double.Remote(transaction, getIID(), isInferred(), type, value);
            }

            @Override
            public final java.lang.Double getValue() {
                return value;
            }

            @Override
            public AttributeTypeImpl.Double getType() {
                return type;
            }

            @Override
            public final AttributeImpl.Double.Remote asDouble() {
                return this;
            }
        }
    }

    public static class String extends AttributeImpl<java.lang.String> implements Attribute.String {

        private final AttributeTypeImpl.String type;
        private final java.lang.String value;

        String(java.lang.String iid, boolean isInferred, AttributeTypeImpl.String type, java.lang.String value) {
            super(iid, isInferred);
            this.type = type;
            this.value = value;
        }

        public static AttributeImpl.String of(ConceptProto.Thing thingProto) {
            return new AttributeImpl.String(
                    bytesToHexString(thingProto.getIid().toByteArray()),
                    thingProto.getInferred(),
                    AttributeTypeImpl.String.of(thingProto.getType()),
                    thingProto.getValue().getString()
            );
        }

        @Override
        public final AttributeTypeImpl.String getType() {
            return type;
        }

        @Override
        public final java.lang.String getValue() {
            return value;
        }

        @Override
        public final AttributeImpl.String asString() {
            return this;
        }

        @Override
        public AttributeImpl.String.Remote asRemote(TypeDBTransaction transaction) {
            return new AttributeImpl.String.Remote(transaction, getIID(), isInferred(), type, value);
        }

        public static class Remote extends AttributeImpl.Remote<java.lang.String> implements Attribute.String.Remote {

            private final AttributeTypeImpl.String type;
            private final java.lang.String value;

            Remote(TypeDBTransaction transaction, java.lang.String iid, boolean isInferred, AttributeTypeImpl.String type, java.lang.String value) {
                super(transaction, iid, isInferred);
                this.type = type;
                this.value = value;
            }

            @Override
            public Attribute.String.Remote asRemote(TypeDBTransaction transaction) {
                return new AttributeImpl.String.Remote(transaction, getIID(), isInferred(), type, value);
            }

            @Override
            public final java.lang.String getValue() {
                return value;
            }

            @Override
            public AttributeTypeImpl.String getType() {
                return type;
            }

            @Override
            public final AttributeImpl.String.Remote asString() {
                return this;
            }
        }
    }

    public static class DateTime extends AttributeImpl<LocalDateTime> implements Attribute.DateTime {

        private final AttributeTypeImpl.DateTime type;
        private final LocalDateTime value;

        DateTime(java.lang.String iid, boolean isInferred, AttributeTypeImpl.DateTime type, LocalDateTime value) {
            super(iid, isInferred);
            this.type = type;
            this.value = value;
        }

        public static AttributeImpl.DateTime of(ConceptProto.Thing thingProto) {
            return new AttributeImpl.DateTime(
                    bytesToHexString(thingProto.getIid().toByteArray()),
                    thingProto.getInferred(),
                    AttributeTypeImpl.DateTime.of(thingProto.getType()),
                    toLocalDateTime(thingProto.getValue().getDateTime())
            );
        }

        @Override
        public final AttributeTypeImpl.DateTime getType() {
            return type;
        }

        @Override
        public final LocalDateTime getValue() {
            return value;
        }

        @Override
        public final AttributeImpl.DateTime asDateTime() {
            return this;
        }

        @Override
        public AttributeImpl.DateTime.Remote asRemote(TypeDBTransaction transaction) {
            return new AttributeImpl.DateTime.Remote(transaction, getIID(), isInferred(), type, value);
        }

        private static LocalDateTime toLocalDateTime(long rpcDatetime) {
            return LocalDateTime.ofInstant(Instant.ofEpochMilli(rpcDatetime), ZoneOffset.UTC);
        }

        public static class Remote extends AttributeImpl.Remote<LocalDateTime> implements Attribute.DateTime.Remote {

            private final AttributeTypeImpl.DateTime type;
            private final LocalDateTime value;

            Remote(TypeDBTransaction transaction, java.lang.String iid, boolean isInferred, AttributeTypeImpl.DateTime type, LocalDateTime value) {
                super(transaction, iid, isInferred);
                this.type = type;
                this.value = value;
            }

            @Override
            public Attribute.DateTime.Remote asRemote(TypeDBTransaction transaction) {
                return new AttributeImpl.DateTime.Remote(transaction, getIID(), isInferred(), type, value);
            }

            @Override
            public final LocalDateTime getValue() {
                return value;
            }

            @Override
            public AttributeTypeImpl.DateTime getType() {
                return type;
            }

            @Override
            public final AttributeImpl.DateTime.Remote asDateTime() {
                return this;
            }
        }
    }
}

File: concept/thing/AttributeImpl.ts
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { AttributeType as AttributeTypeProto, Thing as ThingProto } from "typedb-protocol/common/concept_pb";
import { Attribute } from "../../api/concept/thing/Attribute";
import { Thing } from "../../api/concept/thing/Thing";
import { AttributeType } from "../../api/concept/type/AttributeType";
import { ThingType } from "../../api/concept/type/ThingType";
import { TypeDBTransaction } from "../../api/connection/TypeDBTransaction";
import { ErrorMessage } from "../../common/errors/ErrorMessage";
import { TypeDBClientError } from "../../common/errors/TypeDBClientError";
import { RequestBuilder } from "../../common/rpc/RequestBuilder";
import { Bytes } from "../../common/util/Bytes";
import { Stream } from "../../common/util/Stream";
import { AttributeTypeImpl, ThingImpl } from "../../dependencies_internal";
import BAD_VALUE_TYPE = ErrorMessage.Concept.BAD_VALUE_TYPE;
import INVALID_CONCEPT_CASTING = ErrorMessage.Concept.INVALID_CONCEPT_CASTING;

export abstract class AttributeImpl extends ThingImpl implements Attribute {

    private readonly _type: AttributeType;

    protected constructor(iid: string, inferred: boolean, type: AttributeType) {
        super(iid, inferred);
        this._type = type;
    }

    abstract asRemote(transaction: TypeDBTransaction): Attribute.Remote;

    isAttribute(): boolean {
        return true;
    }

    get type(): AttributeType {
        return this._type;
    }

    abstract get value(): boolean | string | number | Date;

    isBoolean(): boolean {
        return false;
    }

    isDateTime(): boolean {
        return false;
    }

    isDouble(): boolean {
        return false;
    }

    isLong(): boolean {
        return false;
    }

    isString(): boolean {
        return false;
    }

    asAttribute(): Attribute {
        return this;
    }

    JSON(): Record<string, boolean | string | number> {
        let value;
        if (this.value instanceof Date) value = this.value.toISOString().slice(0, -1);
        else value = this.value;
        return {
            type: this.type.label.name,
            value_type: this.type.valueType.name(),
            value: value
        };
    }

    asBoolean(): Attribute.Boolean {
        throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Attribute.Boolean"));
    }

    asLong(): Attribute.Long {
        throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Attribute.Long"));
    }

    asDouble(): Attribute.Double {
        throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Attribute.Double"));
    }

    asString(): Attribute.String {
        throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Attribute.String"));
    }

    asDateTime(): Attribute.DateTime {
        throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Attribute.DateTime"));
    }
}

export namespace AttributeImpl {

    export function of(thingProto: ThingProto): Attribute {
        if (!thingProto) return null;
        const attrType = AttributeTypeImpl.of(thingProto.getType());
        const iid = Bytes.bytesToHexString(thingProto.getIid_asU8());
        const inferred = thingProto.getInferred();
        switch (thingProto.getType().getValueType()) {
            case AttributeTypeProto.ValueType.BOOLEAN:
                return new AttributeImpl.Boolean(iid, inferred, attrType.asBoolean(), thingProto.getValue().getBoolean());
            case AttributeTypeProto.ValueType.LONG:
                return new AttributeImpl.Long(iid, inferred, attrType.asLong(), thingProto.getValue().getLong());
            case AttributeTypeProto.ValueType.DOUBLE:
                return new AttributeImpl.Double(iid, inferred, attrType.asDouble(), thingProto.getValue().getDouble());
            case AttributeTypeProto.ValueType.STRING:
                return new AttributeImpl.String(iid, inferred, attrType.asString(), thingProto.getValue().getString());
            case AttributeTypeProto.ValueType.DATETIME:
                return new AttributeImpl.DateTime(iid, inferred, attrType.asDateTime(), new Date(thingProto.getValue().getDateTime()));
            default:
                throw new TypeDBClientError(BAD_VALUE_TYPE.message(thingProto.getType().getValueType()));
        }
    }

    export abstract class Remote extends ThingImpl.Remote implements Attribute.Remote {

        private readonly _type: AttributeType;

        protected constructor(transaction: TypeDBTransaction.Extended, iid: string, inferred: boolean, type: AttributeType, ..._: any) {
            super(transaction, iid, inferred);
            this._type = type;
        }

        abstract asRemote(transaction: TypeDBTransaction): Attribute.Remote;

        get type(): AttributeType {
            return this._type;
        }

        abstract get value(): boolean | string | number | Date;

        getOwners(ownerType?: ThingType): Stream<Thing> {
            let request;
            if (!ownerType) {
                request = RequestBuilder.Thing.Attribute.getOwnersReq(this.iid);
            } else {
                request = RequestBuilder.Thing.Attribute.getOwnersByTypeReq(this.iid, ThingType.proto(ownerType));
            }
            return this.stream(request)
                .flatMap((resPart) => Stream.array(resPart.getAttributeGetOwnersResPart().getThingsList()))
                .map((thingProto) => ThingImpl.of(thingProto));
        }

        isAttribute(): boolean {
            return true;
        }

        isBoolean(): boolean {
            return false;
        }

        isDateTime(): boolean {
            return false;
        }

        isDouble(): boolean {
            return false;
        }

        isLong(): boolean {
            return false;
        }

        isString(): boolean {
            return false;
        }

        asAttribute(): Attribute.Remote {
            return this;
        }

        JSON(): Record<string, boolean | string | number> {
            let value;
            if (this.value instanceof Date) value = this.value.toISOString().slice(0, -1);
            else value = this.value;
            return {
                type: this.type.label.name,
                value_type: this.type.valueType.name(),
                value: value
            };
        }

        asBoolean(): Attribute.Boolean.Remote {
            throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Attribute.Boolean"));
        }

        asLong(): Attribute.Long.Remote {
            throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Attribute.Long"));
        }

        asDouble(): Attribute.Double.Remote {
            throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Attribute.Double"));
        }

        asString(): Attribute.String.Remote {
            throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Attribute.String"));
        }

        asDateTime(): Attribute.DateTime.Remote {
            throw new TypeDBClientError(INVALID_CONCEPT_CASTING.message(this.className, "Attribute.DateTime"));
        }
    }

    export class Boolean extends AttributeImpl implements Attribute.Boolean {
        private readonly _value: boolean;

        constructor(iid: string, inferred: boolean, type: AttributeType.Boolean, value: boolean) {
            super(iid, inferred, type);
            this._value = value;
        }

        protected get className(): string {
            return "Attribute.Boolean";
        }

        asRemote(transaction: TypeDBTransaction): Attribute.Boolean.Remote {
            return new AttributeImpl.Boolean.Remote(transaction as TypeDBTransaction.Extended, this.iid, this.inferred, this.type, this.value);
        }

        get type(): AttributeType.Boolean {
            return super.type.asBoolean();
        }

        get value(): boolean {
            return this._value;
        }

        isBoolean(): boolean {
            return true;
        }

        asBoolean(): Attribute.Boolean {
            return this;
        }
    }

    export namespace Boolean {

        export class Remote extends AttributeImpl.Remote implements Attribute.Boolean.Remote {

            private readonly _value: boolean;

            constructor(transaction: TypeDBTransaction.Extended, iid: string, inferred: boolean, type: AttributeType.Boolean, value: boolean) {
                super(transaction, iid, inferred, type);
                this._value = value;
            }

            protected get className(): string {
                return "Attribute.Boolean";
            }

            asRemote(transaction: TypeDBTransaction): Attribute.Boolean.Remote {
                return new AttributeImpl.Boolean.Remote(transaction as TypeDBTransaction.Extended, this.iid, this.inferred, this.type, this.value);
            }

            get type(): AttributeType.Boolean {
                return super.type.asBoolean();
            }

            get value(): boolean {
                return this._value;
            }

            isBoolean(): boolean {
                return true;
            }

            asBoolean(): Attribute.Boolean.Remote {
                return this;
            }
        }
    }

    export class Long extends AttributeImpl implements Attribute.Long {
        private readonly _value: number;

        constructor(iid: string, inferred: boolean, type: AttributeType.Long, value: number) {
            super(iid, inferred, type);
            this._value = value;
        }

        protected get className(): string {
            return "Attribute.Long";
        }

        asRemote(transaction: TypeDBTransaction): Attribute.Long.Remote {
            return new AttributeImpl.Long.Remote(transaction as TypeDBTransaction.Extended, this.iid, this.inferred, this.type, this.value);
        }

        get type(): AttributeType.Long {
            return super.type.asLong();
        }

        get value(): number {
            return this._value;
        }

        isLong(): boolean {
            return true;
        }

        asLong(): Attribute.Long {
            return this;
        }
    }

    export namespace Long {

        export class Remote extends AttributeImpl.Remote implements Attribute.Long.Remote {
            private readonly _value: number;

            constructor(transaction: TypeDBTransaction.Extended, iid: string, inferred: boolean, type: AttributeType.Long, value: number) {
                super(transaction, iid, inferred, type);
                this._value = value;
            }

            protected get className(): string {
                return "Attribute.Long";
            }

            asRemote(transaction: TypeDBTransaction): Attribute.Long.Remote {
                return new AttributeImpl.Long.Remote(transaction as TypeDBTransaction.Extended, this.iid, this.inferred, this.type, this.value);
            }

            get type(): AttributeType.Long {
                return super.type.asLong();
            }

            get value(): number {
                return this._value;
            }

            isLong(): boolean {
                return true;
            }

            asLong(): Attribute.Long.Remote {
                return this;
            }
        }
    }

    export class Double extends AttributeImpl implements Attribute.Double {
        private readonly _value: number;

        constructor(iid: string, inferred: boolean, type: AttributeType.Double, value: number) {
            super(iid, inferred, type);
            this._value = value;
        }

        protected get className(): string {
            return "Attribute.Double";
        }

        asRemote(transaction: TypeDBTransaction): Attribute.Double.Remote {
            return new AttributeImpl.Double.Remote(transaction as TypeDBTransaction.Extended, this.iid, this.inferred, this.type, this.value);
        }

        get type(): AttributeType.Double {
            return super.type.asDouble();
        }

        get value(): number {
            return this._value;
        }

        isDouble(): boolean {
            return true;
        }

        asDouble(): Attribute.Double {
            return this;
        }
    }

    export namespace Double {

        export class Remote extends AttributeImpl.Remote implements Attribute.Double.Remote {
            private readonly _value: number;

            constructor(transaction: TypeDBTransaction.Extended, iid: string, inferred: boolean, type: AttributeType.Double, value: number) {
                super(transaction, iid, inferred, type);
                this._value = value;
            }

            protected get className(): string {
                return "Attribute.Double";
            }

            asRemote(transaction: TypeDBTransaction): Attribute.Double.Remote {
                return new AttributeImpl.Double.Remote(transaction as TypeDBTransaction.Extended, this.iid, this.inferred, this.type, this.value);
            }

            get type(): AttributeType.Double {
                return super.type.asDouble();
            }

            get value(): number {
                return this._value;
            }

            isDouble(): boolean {
                return true;
            }

            asDouble(): Attribute.Double.Remote {
                return this;
            }
        }
    }

    export class String extends AttributeImpl implements Attribute.String {
        private _value: string;

        constructor(iid: string, inferred: boolean, type: AttributeType.String, value: string) {
            super(iid, inferred, type);
            this._value = value;
        }

        protected get className(): string {
            return "Attribute.String";
        }

        asRemote(transaction: TypeDBTransaction): Attribute.String.Remote {
            return new AttributeImpl.String.Remote(transaction as TypeDBTransaction.Extended, this.iid, this.inferred, this.type, this.value);
        }

        get type(): AttributeType.String {
            return super.type.asString();
        }

        get value(): string {
            return this._value;
        }

        isString(): boolean {
            return true;
        }

        asString(): Attribute.String {
            return this;
        }
    }

    export namespace String {

        export class Remote extends AttributeImpl.Remote implements Attribute.String.Remote {
            private readonly _value: string;

            constructor(transaction: TypeDBTransaction.Extended, iid: string, inferred: boolean, type: AttributeType.String, value: string) {
                super(transaction, iid, inferred, type);
                this._value = value;
            }

            protected get className(): string {
                return "Attribute.String";
            }

            asRemote(transaction: TypeDBTransaction): Attribute.String.Remote {
                return new AttributeImpl.String.Remote(transaction as TypeDBTransaction.Extended, this.iid, this.inferred, this.type, this.value);
            }

            get type(): AttributeType.String {
                return super.type.asString();
            }

            get value(): string {
                return this._value;
            }

            isString(): boolean {
                return true;
            }

            asString(): Attribute.String.Remote {
                return this;
            }
        }
    }

    export class DateTime extends AttributeImpl implements Attribute.DateTime {
        private readonly _value: Date;

        constructor(iid: string, inferred: boolean, type: AttributeType.DateTime, value: Date) {
            super(iid, inferred, type);
            this._value = value;
        }

        protected get className(): string {
            return "Attribute.DateTime";
        }

        asRemote(transaction: TypeDBTransaction): Attribute.DateTime.Remote {
            return new AttributeImpl.DateTime.Remote(transaction as TypeDBTransaction.Extended, this.iid, this.inferred, this.type, this.value);
        }

        get type(): AttributeType.DateTime {
            return super.type.asDateTime();
        }

        get value(): Date {
            return this._value;
        }

        isDateTime(): boolean {
            return true;
        }

        asDateTime(): Attribute.DateTime {
            return this;
        }
    }

    export namespace DateTime {

        export class Remote extends AttributeImpl.Remote implements Attribute.DateTime.Remote {
            private readonly _value: Date;

            constructor(transaction: TypeDBTransaction.Extended, iid: string, inferred: boolean, type: AttributeType.DateTime, value: Date) {
                super(transaction, iid, inferred, type);
                this._value = value;
            }

            protected get className(): string {
                return "Attribute.DateTime.Remote";
            }

            asRemote(transaction: TypeDBTransaction): Attribute.DateTime.Remote {
                return new AttributeImpl.DateTime.Remote(transaction as TypeDBTransaction.Extended, this.iid, this.inferred, this.type, this.value);
            }

            get type(): AttributeType.DateTime {
                return super.type.asDateTime();
            }

            get value(): Date {
                return this._value;
            }

            isDateTime(): boolean {
                return true;
            }

            asDateTime(): Attribute.DateTime.Remote {
                return this;
            }
        }
    }
}
